nengo_viz/components/netgraph.py

```diff
@@ -21,19 +23,128 @@ def __init__(self, viz, config, uid):
+        self.initialized_pan_and_zoom = False
+        self.last_modify_time = os.path.getmtime(self.viz.viz.filename)
 
-    def get_parents(self, uid):
+    def check_for_reload(self):
+        t = os.path.getmtime(self.viz.viz.filename)
+
+        if self.last_modify_time is None or self.last_modify_time < t:
+            self.reload()
+            self.last_modify_time = t
+
+    def reload(self):
+        locals = {}
+        with open(self.viz.viz.filename) as f:
+            code = f.read()
+        try:
+            exec(code, locals)
+        except:
+            traceback.print_exc()
+        try:
+            model = locals['model']
+        except:
+            traceback.print_exc()
+            return
+
+        locals['nengo_viz'] = nengo_viz
+        name_finder = nengo_viz.NameFinder(locals, model)
+
+        self.networks_to_search = [model]
+        self.parents = {}
+
+        for uid, old_item in self.uids.items():
+            try:
+                new_item = eval(uid, locals)
+            except:
+                new_item = None
+
+            if new_item is None:
+                self.to_be_sent.append(dict(
+                    type='remove', uid=uid))
+                del self.uids[uid]
+            else:
+                if isinstance(old_item, (nengo.Node,
+                                         nengo.Ensemble,
+                                         nengo.Network)):
+                    old_label = self.viz.viz.get_label(old_item)
+                    new_label = self.viz.viz.get_label(new_item,
+                        default_labels=name_finder.known_name)
+
+                    if old_label != new_label:
+                        self.to_be_sent.append(dict(
+                            type='rename', uid=uid, name=new_label))
+                elif isinstance(old_item, nengo.Connection):
+                    old_pre = old_item.pre_obj
+                    old_post = old_item.post_obj
+                    new_pre = new_item.pre_obj
+                    new_post = new_item.post_obj
+                    if isinstance(old_pre, nengo.ensemble.Neurons):
+                        old_pre = old_pre.ensemble
+                    if isinstance(old_post, nengo.ensemble.Neurons):
+                        old_post = old_post.ensemble
+                    if isinstance(new_pre, nengo.ensemble.Neurons):
+                        new_pre = new_pre.ensemble
+                    if isinstance(new_post, nengo.ensemble.Neurons):
+                        new_post = new_post.ensemble
+
+                    old_pre = self.viz.viz.get_uid(old_pre)
+                    old_post = self.viz.viz.get_uid(old_post)
+                    new_pre = self.viz.viz.get_uid(new_pre,
+                            default_labels=name_finder.known_name)
+                    new_post = self.viz.viz.get_uid(new_post,
+                            default_labels=name_finder.known_name)
+
+                    if new_pre != old_pre or new_post != old_post:
```

Comment:
Nope, this is checking whether the objects are the "same" in the sense of they have the same uid.  They won't be the same object, since when the code changes, we revaluate the script, which creates all-new objects.  So we use the uid to figure out object identity

This gets complicated for Connections, since they don't really have a sensible uid.  (something like ```model.connections[3]``` is about as close as we could get to a uid, but then that breaks horribly when someone adds or removes a connection).  So I'm using the uid of the pre and the post of the connection instead.